spec/types/branded-string.ts

```diff
@@ -0,0 +1,13 @@
+import * as Spec from './spec'
+import { Brand, Type } from './typebox'
+
+type Email = Brand<string, 'Email'>
+type WrongEmail = Brand<string, 'WrongEmail'>
+
+// For branded types we require to pass extra constraints, otherwise it would
+// make no sense to use them at all.
+Spec.expectType<Email>(Spec.infer(Type.BrandedString<Email>({ format: 'email' })))
+
+// We validate that brands clearly distinguish between different types.
+Spec.expectNotType<Email>(Spec.infer(Type.BrandedString<WrongEmail>({ format: 'email' })))
+Spec.expectNotType<Email>(Spec.infer(Type.String()))
```

Comment:
This showcases the introduced feature:
  - brand strings only work when we pass "configuration" to `BrandedString` (otherwise it would make no sense to attach brands to those values)
  - I had to use a different function than `String` to avoid breaking backwards compatibility. The name is completely debatable, it doesn't have to be `BrandedString`, I'm open to suggestions.

Reply:
I've actually been thinking a little bit about what would constitute a nominal type in JSON Schema form. Typically types in nominal type systems are derived from their simple name, so for example the following two types would be seen as distinct in such a system, in JSON schema terms, their names would be implied by their `$id` properties (which is the closest JSON schema has to naming a type)

```typescript
const Vector = Type.Object({
   x: Type.Number(),
   y: Type.Number(),
   z: Type.Number()
}, { $id: 'Vector' }) // $id === brand

const AnotherVector = Type.Object({
   x: Type.Number(),
   y: Type.Number(),
   z: Type.Number()
}, { $id: 'AnotherVector ' }) // $id === brand
```
Unfortunately, both TypeScript and JSON Schema would see these types as one in the same, irrespective of the `$id`, the  types themselves are structurally equivalent. I think for the `Type.BrandedType({...})`, the `Brand<string, 'Email'>`, the `Email` literal would be inferred from the types `$id` property. Such that.

```typescript
const Email = Type.String({ format: 'email', $id: 'Email' })
```